### PR TITLE
Improve Webcrypto error handling, resolve several internal errors and add clarifying comments

### PR DESCRIPTION
**Pull request by fhanau**
Commit messages and comments should explain the changes reasonably well, please let me know if you find anything unclear.
One issue that merits discussion is whether to validate imported RSA keys. Calling validateRsaParams() should only reject keys that are unreasonably large or possibly insecure based on the modulus size or exponent, but may break existing scripts that import such keys. Fortunately, the other changes add some key size checks during signing and should be sufficient to prevent internal errors even if unreasonably small keys are imported.
For signing with RSASSA-PKCS1-v1_5, in very few cases errors will be reported where there previously was no error (e.g. when importing a key of exactly 62 or 63 bytes, using SHA256 as the hash digest and trying to use it for signing). I think the change – requiring the key to be at least 32 bytes larger than the hash digest when 26 or 30 would be sufficient depending on which hash is used – is sensible since there is no good reason to use keys of these specific sizes, boringssl doesn't generate keys of that size already and it is unlikely anyone is using this exact configuration.

### CODE REVIEW

File: src/workerd/api/crypto-impl-asymmetric.c++ [hunk 1]
```diff
@@ -941,6 +980,9 @@ kj::Own<CryptoKey::Impl> CryptoKey::Impl::importRsaRaw(
   auto publicExponent = kj::heapArray<kj::byte>(BN_num_bytes(e));
   KJ_ASSERT(BN_bn2bin(e, publicExponent.begin()) == publicExponent.size());
 
+  //Validate modulus and exponent, reject imported RSA keys that may be unsafe.
```
Comment (jasnell) [inline src/workerd/api/crypto-impl-asymmetric.c++:983]:
```suggestion
  // Validate modulus and exponent, reject imported RSA keys that may be unsafe.
```

File: src/workerd/api/crypto-impl-asymmetric.c++ [hunk 2]
```diff
@@ -1438,6 +1480,7 @@ kj::Own<EVP_PKEY> ellipticJwkReader(int curveId, SubtleCrypto::JsonWebKey keyDat
 ImportAsymmetricResult importEllipticRaw(SubtleCrypto::ImportKeyData keyData, int curveId,
     kj::StringPtr normalizedName, kj::ArrayPtr<const kj::String> keyUsages,
     CryptoKeyUsageSet allowedUsages) {
+  //Import an elliptic key represented by raw data, only public keys are supported.
```
Comment (jasnell) [inline src/workerd/api/crypto-impl-asymmetric.c++:1483]:
```suggestion
  // Import an elliptic key represented by raw data, only public keys are supported.
```

File: src/workerd/api/crypto-impl-asymmetric.c++ [hunk 3]
```diff
@@ -875,6 +909,9 @@ kj::Own<CryptoKey::Impl> CryptoKey::Impl::importRsa(
   auto publicExponent = kj::heapArray<kj::byte>(BN_num_bytes(e));
   KJ_ASSERT(BN_bn2bin(e, publicExponent.begin()) == publicExponent.size());
 
+  //Validate modulus and exponent, reject imported RSA keys that may be unsafe.
```
Comment (jasnell) [inline src/workerd/api/crypto-impl-asymmetric.c++:912]:
```suggestion
  // Validate modulus and exponent, reject imported RSA keys that may be unsafe.
```

File: src/workerd/api/crypto-impl-asymmetric.c++ [hunk 4]
```diff
@@ -702,6 +730,12 @@ kj::OneOf<jsg::Ref<CryptoKey>, CryptoKeyPair> CryptoKey::Impl::generateRsa(
       publicExponent.size(), nullptr), InternalDOMOperationError, "Error setting up RSA keygen.");
 
   auto rsaPrivateKey = OSSL_NEW(RSA);
+  // TODO: boringssl silently uses (modulusLength & ~127) for the key size, i.e. it rounds down to
```
Comment (jasnell) [inline src/workerd/api/crypto-impl-asymmetric.c++:733]: Nit: We'll typically use a style like `TODO(soon)` or `TODO(conform)` etc. Non-blocking tho

File: src/workerd/api/crypto-impl-asymmetric.c++ [hunk 5]
```diff
@@ -103,6 +103,31 @@ public:
...
+      RSA& rsa = JSG_REQUIRE_NONNULL(EVP_PKEY_get0_RSA(getEvpPkey()), DOMDataError,
+          "Missing RSA key", tryDescribeOpensslErrors());
+      auto salt = JSG_REQUIRE_NONNULL(algorithm.saltLength, TypeError,
+          "Failed to provide salt for RSA-PSS key operation which requires a salt");
+      JSG_REQUIRE(salt >= 0, TypeError, "SaltLength for RSA-PSS must be non-negative (provided ",
+          salt, ").");
+      JSG_REQUIRE(EVP_MD_size(type) + 2 <= RSA_size(&rsa), DOMInvalidAccessError,
+          "key too small for signing with given digest");
+      JSG_REQUIRE(salt <= RSA_size(&rsa) - EVP_MD_size(type) - 2, DOMInvalidAccessError,
+          "key too small for signing with given digest and salt length");
```
Comment (jasnell) [inline src/workerd/api/crypto-impl-asymmetric.c++:129]: Let's make these all `DOMException`s for consistency.

**Comment by jasnell** [comment]
Needs an internal PR to run through CI.

File: src/workerd/api/crypto-impl-asymmetric.c++ [hunk 6]
```diff
@@ -730,6 +730,12 @@ kj::OneOf<jsg::Ref<CryptoKey>, CryptoKeyPair> CryptoKey::Impl::generateRsa(
       publicExponent.size(), nullptr), InternalDOMOperationError, "Error setting up RSA keygen.");
 
   auto rsaPrivateKey = OSSL_NEW(RSA);
+  // TODO: boringssl silently uses (modulusLength & ~127) for the key size, i.e. it rounds down to
+  // the closest multiple of 128 bits. This can easily cause confusion when non-standard key sizes
+  // are requested. Ideally we would throw an error when trying to create keys where the size would
+  // be rounded down, but this would likely break existing scripts.
```
Comment (harrishancock) [inline src/workerd/api/crypto-impl-asymmetric.c++:736]: We could emit a log message to confirm whether we'd break anyone, like here: https://github.com/cloudflare/workerd/blob/0efe8aeee69d1a0ae3939fa0ccec15dd758ccd3a/src/workerd/io/worker.c%2B%2B#L377-L381

Comment (harrishancock) [reply src/workerd/api/crypto-impl-asymmetric.c++:736]: We could also do the same for the other cases where you identify that we could theoretically be introducing a breaking change, e.g. when calling validateRsaParams() during RSA key import, and 62/63-byte RSASSA-PKCS1-v1_5 keys used with SHA-256.

**Comment by fhanau** [comment]
> Needs an internal PR to run through CI.

There is an internal branch (felix/crypto-errors2), would that generally be enough or is a PR preferred? The internal tests seem to be failing for some reason though – I'll take care of the suggested changes, rebase and see if that fixes it.

File: src/workerd/api/crypto-impl-asymmetric.c++ (hunk 6 above)
Comment (fhanau) [reply src/workerd/api/crypto-impl-asymmetric.c++:736]: Sounds good, I'll try doing that. Can you explain why there's a lambda expression around this instance of KJ_LOG – is it related to how the warning is reported?

**Comment by jasnell** [comment]
A branch is fine so long as CI shows that internal changes aren't also needed for this to land, and that nothing breaks there of course :-)

File: src/workerd/api/crypto-impl-asymmetric.c++ (hunk 6 above)
Comment (fhanau) [reply src/workerd/api/crypto-impl-asymmetric.c++]: Switched to log messages to detect potential breakage for key signing and validating imported keys. For the key size issue I'd expect that there likely are scripts using sizes other than multiples of 128, so I left the TODO message as-is for now.

File: src/workerd/api/crypto-impl-asymmetric.c++ (hunk 5 above)
Comment (jasnell) [reply src/workerd/api/crypto-impl-asymmetric.c++:157]: Not sure if you meant to mark this one resolved. There's still a `TypeError` there on line 152 that should be a `DOMException`

Comment (fhanau) [reply src/workerd/api/crypto-impl-asymmetric.c++:157]: Sorry, I misinterpreted your comment. Should be fixed now

**Comment by fhanau** [comment]
Internal branch felix/crypto-errors2 now passes all test cases.

File: src/workerd/api/crypto-impl-asymmetric.c++ [hunk 7]
```diff
@@ -103,6 +103,59 @@ public:
...
 
     auto type = lookupDigestAlgorithm(chooseHash(algorithm.hash)).second;
+    if (getAlgorithmName() == "RSASSA-PKCS1-v1_5") {
+      // RSASSA-PKCS1-v1_5 requires the RSA key to be at least as big as the digest size
+      // plus a 15 to 19 byte digest-specific prefix (see boringssl's RSA_add_pkcs1_prefix) plus 11
+      // bytes for padding (see RSA_PKCS1_PADDING_SIZE). For simplicity, require the key to be at
+      // least 32 bytes larger than the hash digest.
+      // Similar checks could also be adopted for more detailed error handling in verify(), but the
+      // current approach should be sufficient to avoid internal errors.
+      RSA& rsa = JSG_REQUIRE_NONNULL(EVP_PKEY_get0_RSA(getEvpPkey()), DOMDataError,
```
Comment (jasnell) [inline src/workerd/api/crypto-impl-asymmetric.c++:113]: The only remaining question here is whether it's possible that adding the additional errors here could break any existing code. If so, these *might* need to go behind a compatibility flag. If these are cases that simply wouldn't have worked correctly previously (causing an internal error as opposed to a useful DOMException) then cool, no compat flag needed. If, however, these cases previously would not have thrown, adding the error could break someone.

File: src/workerd/api/crypto-impl-asymmetric.c++ [hunk 8]
```diff
@@ -644,12 +711,36 @@ kj::Maybe<T> fromBignum(kj::ArrayPtr<kj::byte> value) {
...
+        KJ_LOG(WARNING, "Imported RSA key has invalid modulus length ", modulusLength, ".");
+        return true;
+      })();
+    }
+    KJ_IF_MAYBE(v, fromBignum<unsigned>(publicExponent)) {
+      if (*v != 3 && *v != 65537) {
+        static bool logOnce KJ_UNUSED = ([v] {
+          KJ_LOG(WARNING, "Imported RSA key has invalid publicExponent ", *v,".");
+          return true;
+        })();
```
Comment (jasnell) [inline src/workerd/api/crypto-impl-asymmetric.c++:735]: No action for this PR but just noting... this is a common enough pattern now that we could probably justify adding a new `LOG_WARNING_ONCE("...")` macro to simplify.

Comment (fhanau) [reply src/workerd/api/crypto-impl-asymmetric.c++:735]: That's a good suggestion! Looks like this approach is not used that much elsewhere, but if I use it again in the future I'll try adding a macro

File: src/workerd/api/crypto-impl-asymmetric.c++ (hunk 7 above)
Comment (fhanau) [reply src/workerd/api/crypto-impl-asymmetric.c++:113]: Overall, I'm confident that after switching to just logging warnings in some cases, the additional error statements will only produce errors when there previously would have been an internal error/a different error message. In this function the math gets relatively complex but based on boringssl's code and my tests it should be correct.

**Review by jasnell — APPROVED** [review]

**Comment by panva** [comment]
While there are some eyes on subtle crypto 🙏  #138, #48, and of course at least Ed25519 from https://wicg.github.io/webcrypto-secure-curves/

**Comment by fhanau** [comment]
@panva I'll work on the secure curves and take a look at the other issues in the coming days.

**Pull request merged**